tests/model_validation/tests.py

```diff
@@ -29,28 +21,64 @@ def test_models_validate(self):
         #       See: https://code.djangoproject.com/ticket/21375
         management.call_command("check", stdout=six.StringIO())
 
-    def test_model_signal(self):
-        unresolved_references = post_init.unresolved_references.copy()
-        post_init.connect(on_post_init, sender='missing-app.Model')
-        post_init.connect(OnPostInit(), sender='missing-app.Model')
+    @isolate_apps('django.contrib.auth', kwarg_name='apps')
+    def test_lazy_reference_checks(self, apps):
+
+        class DummyClass(object):
+            def __call__(self, **kwargs):
+                pass
 
-        errors = run_checks()
+            def dummy_method(self):
+                pass
+
+        def dummy_function(*args, **kwargs):
+            pass
+
+        apps.lazy_model_operation(dummy_function, ('auth', 'imaginarymodel'))
+        apps.lazy_model_operation(dummy_function, ('fanciful_app', 'imaginarymodel'))
+
+        post_init.connect(dummy_function, sender='missing-app.Model', apps=apps)
+        post_init.connect(DummyClass(), sender='missing-app.Model', apps=apps)
+        post_init.connect(DummyClass().dummy_method, sender='missing-app.Model', apps=apps)
+
+        errors = _check_lazy_references(apps)
         expected = [
             Error(
-                "The 'on_post_init' function was connected to the 'post_init' "
-                "signal with a lazy reference to the 'missing-app.Model' "
-                "sender, which has not been installed.",
+                "%s contains a lazy reference to auth.imaginarymodel, "
+                "but app 'auth' doesn't provide model 'imaginarymodel'."
+                % repr(dummy_function),
```

Comment:
Latest edits look good. Just noticed a small thing: is there any problem using `%r` instead of `%s` and `repr()`?